File: src/API/Dtos/System/FunctionVM.cs
namespace API.Dtos.System;

public class FunctionVM
{
    public string Id { get; set; } = string.Empty;

    public string Name { get; set; } = string.Empty;

    public string Url { get; set; } = string.Empty;

    public int SortOrder { get; set; }

    public string ParentId { get; set; } = string.Empty;

    public string Icon { get; set; } = string.Empty;
    public List<string> CommandInFunction { get; set; } = [];
}

public interface ITreeNode
{
    string Id { get; set; }
    string Name { get; set; }
    public string Url { get; set; }

    public int SortOrder { get; set; }

    public string? ParentId { get; set; }

    public string Icon { get; set; }
    List<ITreeNode> Children { get; set; }
}

public class FunctionTreeVM : ITreeNode
{
    public string Id { get; set; } = string.Empty;

    public string Name { get; set; } = string.Empty;

    public string Url { get; set; } = string.Empty;

    public int SortOrder { get; set; }

    public string? ParentId { get; set; } = string.Empty;

    public string Icon { get; set; } = string.Empty;
    public List<ITreeNode> Children { get; set; } = [];
}

File: src/SPA/src/app/_core/models/system/functionvm.ts
// using for menu bar left
export interface FunctionVM {
  id: string;
  name: string;
  url: string;
  sortOrder: number;
  parentId: string;
  icon: string;
  commandInFunction: string[];
  children?: FunctionVM[];
  newLinkFlag: boolean;
  open?: boolean;
  selected?: boolean; // Check or not
  menuType: 'C' | 'F'; // c: menu, f button
}

export interface ITreeNode {
  id: string;
  parentId: string | null;
  children: ITreeNode[];
}

export interface FunctionTreeVM {
  id: string;
  name: string;
  url: string;
  sortOrder: number;
  parentId: string | null;
  icon: string;
  newLinkFlag: boolean;
  open?: boolean;
  selected?: boolean; // Check or not
  menuType: 'C' | 'F'; // c: menu, f button
  children: FunctionTreeVM[];
}
